### PR TITLE
TitlePart required, AliasPart disabled

### LINKED ISSUE #5036: Make AliasPart form controls hide like TitlePart
image

You can use Liquid, so it's as good as TitlePart.

### PR DESCRIPTION
**Pull request by PiemP**
Fixes #5036, #5315

- TitlePartSettings, TitlePartSettingsViewModel: Added a boolean property to manage the "Required" setting
- TitlePartSettings.Edit.cshtml : added a checkbox for the "Required" field
- TitlePartDisplay.cs :  added check to avoid empty value, if title is required
- AliasPart: added 2 editors to display  it as hidden input or display it as disabled input
- AliasPart: added language direction support
- AliasPartDisplayDriver.cs : added GetEditorShapeType call to allow multiple editor type

Waiting for your feedback.

### CODE REVIEW

**Comment by deanmarcussen** [comment]
Couple of thoughts.

The required for Title makes good sense. 

But it is not going to be possible to generate a title with a pattern if the title has been set required.

So it might be better to make it part of the enum options.

Maybe `EditableAndRequired` ?

Does that make sense?

Also it seems strange to totally hide the alias part. 
I understand why you would want it disabled. 
But the alias is an identifier part, so if it is hidden, how could you know what the identifier is?

**Comment by PiemP** [comment]
> So it might be better to make it part of the enum options.
> 
> Maybe `EditableAndRequired` ?
> 
> Does that make sense?

I have added the checkbox because is used over all others fields that are required. If you prefer, there is no problem for me to make it an option of the enumerator.

> But the alias is an identifier part, so if it is hidden, how could you know what the identifier is?

I have done it because my boss want to give the possibility to edit a content type to users and he think that see alias could be strange for users, but alias could be useful for me to do some activity over items. For example: my boss want to allow end-users to edit menu and it have alias and I need alias to display it.

**Comment by deanmarcussen** [comment]
> I have added the checkbox because is used over all others fields that are required. If you prefer, there is no problem for me to make it an option of the enumerator.

I think it is better to be a part of the enumerator, because when you have made it required, the liquid template is generated in the handler, but validation will have failed in the driver because it is required, so it would be impossible to have a generated title, that is also required.

> > But the alias is an identifier part, so if it is hidden, how could you know what the identifier is?
> 
> I have done it because my boss want to give the possibility to edit a content type to users and he think that see alias could be strange for users, but alias could be useful for me to do some activity over items. For example: my boss want to allow end-users to edit menu and it have the alias field and I use the alias to display it.
> 
> I make it hidden but if I want to see alias value I can simply see the value in the hidden field.

For this I disagree it being part of OrchardCore, as it could be super confusing for users. And only makes sense because of what your boss asks you to do.
So I would suggest you create the editor option in your own code, to do what you want. (the idea with editors is that you can create your own, as and when required for these special options)

I would also suggest that the better option might be to move fields you don't want editors to see to a tab with placement, rather than hiding them completing. Reading element values from the DOM is not something I think we want to be suggesting to every user to do.

Regarding the disabled editor option. Also this is a bit confusing to have as simply an editor mode. It should be treated the same way as the title settings are currently, as part of an enum.

I am ok with the change to support an editor type for the alias part, just not to actually implement any in the core code base.

**Comment by PiemP** [comment]
I have made the changes. Hope to have understand all the requirements.

File: src/OrchardCore.Modules/OrchardCore.Alias/Settings/AliasPartSettingsDisplayDriver.cs [hunk 1]
```diff
@@ -46,15 +47,15 @@ public override async Task<IDisplayResult> UpdateAsync(ContentTypePartDefinition
 
             var model = new AliasPartSettingsViewModel();
 
-            if (await context.Updater.TryUpdateModelAsync(model, Prefix, m => m.Pattern))
+            if (await context.Updater.TryUpdateModelAsync(model, Prefix, m => m.Pattern, m=>m.Options))
             {
```
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Alias/Settings/AliasPartSettingsDisplayDriver.cs]: formatting here `, m => m.Options`

**Comment by deanmarcussen** [comment]
Oh, can you update the pr title as well

File: src/OrchardCore.Modules/OrchardCore.Alias/Settings/AliasPartSettingsDisplayDriver.cs (hunk 1 above)
Comment (PiemP) [reply src/OrchardCore.Modules/OrchardCore.Alias/Settings/AliasPartSettingsDisplayDriver.cs]: Thank you @deanmarcussen ... I have to remove my subscription to the ugly code society 😅

File: src/OrchardCore.Modules/OrchardCore.Title/Drivers/TitlePartDisplay.cs [hunk 2]
```diff
@@ -41,7 +44,14 @@ public override IDisplayResult Edit(TitlePart titlePart, BuildPartEditorContext
 
         public override async Task<IDisplayResult> UpdateAsync(TitlePart model, IUpdateModel updater, UpdatePartEditorContext context)
         {
-            await updater.TryUpdateModelAsync(model, Prefix, t => t.Title);
+            if(await updater.TryUpdateModelAsync(model, Prefix, t => t.Title))
```
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Title/Drivers/TitlePartDisplay.cs]:
```suggestion
            if (await updater.TryUpdateModelAsync(model, Prefix, t => t.Title))
```

**Pull request merged**